include/boost/random/detail/gray_coded_qrng.hpp

```diff
@@ -0,0 +1,152 @@
+/* boost random/detail/gray_coded_qrng.hpp header file
+ *
+ * Copyright Justinas Vygintas Daugmaudis 2010-2018
+ * Distributed under the Boost Software License, Version 1.0. (See
+ * accompanying file LICENSE_1_0.txt or copy at
+ * http://www.boost.org/LICENSE_1_0.txt)
+ */
+
+#ifndef BOOST_RANDOM_DETAIL_GRAY_CODED_QRNG_HPP
+#define BOOST_RANDOM_DETAIL_GRAY_CODED_QRNG_HPP
+
+#include <boost/random/detail/qrng_base.hpp>
+
+#include <boost/multiprecision/integer.hpp> // lsb
+
+#include <functional> // bit_xor
+
+#include <boost/mpl/if.hpp>
+
+//!\file
+//!Describes the gray-coded quasi-random number generator base class template.
+
+namespace boost {
+namespace random {
+
+namespace detail {
+
+template<typename LatticeT>
+class gray_coded_qrng
+  : public qrng_base<
+      gray_coded_qrng<LatticeT>
+    , LatticeT
+    , typename LatticeT::value_type
+    >
+{
+public:
+  typedef typename LatticeT::value_type result_type;
+  typedef result_type size_type;
+
+private:
+  typedef gray_coded_qrng<LatticeT> self_t;
+  typedef qrng_base<self_t, LatticeT, size_type> base_t;
+
+  // The base needs to access modifying member f-ns, and we
+  // don't want these functions to be available for the public use
+  friend class qrng_base<self_t, LatticeT, size_type>;
+
+  // Respect lattice bit_count here
+  struct do_nothing {
+    inline static void check(unsigned) {}
+    inline static void check_log2(size_type) {}
+  };
+  struct check_bit_range {
+    inline static void check(unsigned bit_pos) {
+      if (bit_pos >= LatticeT::bit_count)
+        boost::throw_exception( std::range_error("gray_coded_qrng: bit_count") );
+    }
+    inline static void check_log2(size_type code) {
+      check_bit_range::check(multiprecision::msb(code));
+    }
+  };
+
+  // We only want to check whether bit pos is outside the range if given bit_count
+  // is narrower than the size_type, otherwise checks compile to nothing.
+  typedef typename mpl::if_c<
+      LatticeT::bit_count < std::numeric_limits<size_type>::digits
+    , check_bit_range
+    , do_nothing
+  >::type bit_range_checker_t;
+
+public:
+  //!Returns: Tight lower bound on the set of values returned by operator().
+  //!
+  //!Throws: nothing.
+  static BOOST_CONSTEXPR result_type min BOOST_PREVENT_MACRO_SUBSTITUTION ()
+  { return 0; }
+
+  //!Returns: Tight upper bound on the set of values returned by operator().
+  //!
+  //!Throws: nothing.
+  static BOOST_CONSTEXPR result_type max BOOST_PREVENT_MACRO_SUBSTITUTION ()
+  { return (std::numeric_limits<result_type>::max)(); }
```

Comment:
This doesn't respect bit_count.  Use boost::low_bits_mask_t (from boost/integer/integer_mask.hpp).

Reply:
Nice idea, but if I implement max() as 
{ return low_bits_mask_t<LatticeT::bit_count>::sig_bits; }

then 8 bit discard test doesn't compile because boost/integer/integer_mask.hpp:66:5: error: left operand of shift expression ‘(-1 << 7)’ is negative [-fpermissive]
     BOOST_STATIC_CONSTANT( least, sig_bits = (~( ~(least( 0u )) << Bits )) );

Basically, least is resolved to char, and it gets negative.

For the time being I would suggest to write this as (std::numeric_limits<result_type>::max)() >> (std::numeric_limits<IntType>::digits - LatticeT::bit_count).

This will produce the result we want until this problem with integer_mask gets resolved.